apps/prairielearn/python/test/core_test.py

```diff
@@ -636,96 +677,574 @@ def test_string_to_2darray_invalid(
+        (
+            np.array([[1.0, 2.0], [3.0, 4.0]]),
+            np.array([[1.0, 2.0], [3.0, 4.0]]),
+            2,
+            True,
+        ),
+        # Within tolerance (2 decimal places = +/- 0.005)
+        (
+            np.array([[1.001, 2.005], [3.008, 4.009]]),
+            np.array([[1.005, 2.000], [3.004, 4.014]]),
+            2,
+            True,
+        ),
+        # Outside tolerance
+        (
+            np.array([[1.01, 2.0], [3.0, 4.0]]),
+            np.array([[1.00, 2.0], [3.0, 4.0]]),
+            2,
+            False,
+        ),
+        # Complex numbers
+        (
+            np.array([[1 + 1j, 2 + 2j], [3 + 3j, 4 + 4j]]),
+            np.array([[1 + 1j, 2 + 2j], [3 + 3j, 4 + 4j]]),
+            2,
+            True,
+        ),
     ],
 )
-def test_clean_identifier_name(*, input_name: str, expected_output: str) -> None:
-    """Test clean_identifier_name with various input strings."""
-    assert pl.clean_identifier_name(input_name) == expected_output
+def test_is_correct_ndarray2d_dd(
+    *,
+    submitted: np.ndarray[Any, Any],
+    true: np.ndarray[Any, Any],
+    digits: int,
+    expected: bool,
+) -> None:
+    """Test arrays are equal within decimal digit tolerance."""
+    assert pl.is_correct_ndarray2d_dd(submitted, true, digits) == expected
+
+
+@pytest.mark.parametrize(
+    ("submitted", "true", "digits", "expected"),
+    [
+        # Basic array equality
+        (
+            np.array([[1.0, np.inf], [3.0, np.nan]]),
+            np.array([[1.0, np.inf], [3.0, np.nan]]),
+            2,
+            True,
+        ),
+        # Within tolerance (2 significant figures)
+        (
+            np.array([[0.095, 2.0], [3.0, 4.0]]),
+            np.array([[0.100, 2.0], [3.0, 4.0]]),
+            2,
+            True,
+        ),
+        # Outside tolerance
+        (
+            np.array([[0.094, 2.0], [3.0, 4.0]]),
+            np.array([[0.100, 2.0], [3.0, 4.0]]),
+            2,
+            False,
+        ),
+        # Complex numbers
+        (
+            np.array([[1 + 1j, 2 + 2j], [3 + 3j, 4 + 4j]]),
+            np.array([[1 + 1j, 2 + 2j], [3 + 3j, 4 + 4j]]),
+            2,
+            True,
+        ),
+        (
+            np.array([[1 + 0.094j, 2 + 2j], [3 + 3j, 4 + 4j]]),
+            np.array([[1 + 1j, 2 + 2j], [3 + 3j, 4 + 4j]]),
+            2,
+            False,
+        ),
+    ],
+)
+def test_is_correct_ndarray2d_sf(
+    *,
+    submitted: np.ndarray[Any, Any],
+    true: np.ndarray[Any, Any],
+    digits: int,
+    expected: bool,
+) -> None:
+    """Test arrays are equal within significant figure tolerance."""
+    assert pl.is_correct_ndarray2d_sf(submitted, true, digits) == expected
+
+
+def test_load_extension() -> None:
+    """Test loading extensions with the load_extension function."""
+    director = Path(__file__).parent
+    controller = "dummy_extension.py"
+
+    # Create mock data with extension info
+    data: pl.QuestionData = {
```

Comment:
I think you can use the question data currently given in the test fixture and add the relevant entries instead of creating a new one from scratch.

Reply:
Where is this defined so that the other test cases work? I'm very confused there

Reply:
https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/python/test/conftest.py